### PR TITLE
Add support for templatePatch

### LINKED ISSUE #368: Add support of templatePatch in ApplicationSet (on ArgoCD 2.10)
### Description

A new parameter named `templatePatch` allow to add more control on Application template in ApplicationSet. More information [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Template/#template-patch)

### Potential Terraform Configuration

```hcl
locals {
  mytemplate = <<EOT
      spec:
      source:
        helm:
          valueFiles:
          {{- range $valueFile := .valueFiles }}
            - {{ $valueFile }}
          {{- end }}
    {{- if .autoSync }}
      syncPolicy:
        automated:
          prune: {{ .prune }}
    {{- end }}
    EOT
  project = "my-project"
}

resource "argocd_application_set" "test" {
  metadata {
    name      = "my-applicationset"
    namespace = "argocd"
  }

  spec {
    generator {
      clusters {
        selector {
          match_labels = {
            "env" = "qa"
          }
        }
      }
    }
    go_template = true
    templatePatch = local.mytemplate
    template {
      metadata {
        name = "test-{{ .name }}"
      }
      spec {
        destination {
          namespace = "my-ns"
          server    = "{{ .server }}"
        }
        project = local.project

        source {
          repo_url        = "{{ .url  }}"
          path            = "{{ .chart  }}"
          target_revision = "{{ .branch  }}"
        }
      }
      source {
        repo_url        = "{{ .url }}"
        ref             = "local"
        target_revision = "{{ .branch }}"
        path            = ""
      }
    }
  }
}

```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by KyriosGN0**
closes #368 
still needs some manual testing

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -25,7 +25,7 @@ require (
 	github.com/stretchr/testify v1.9.0
 	golang.org/x/crypto v0.28.0
 	google.golang.org/protobuf v1.35.1
-	k8s.io/apiextensions-apiserver v0.29.6
+	k8s.io/apiextensions-apiserver v0.31.1
```
Comment (the-technat) [inline go.mod:28]: Do you know why this bump is needed? (Usually we try to keep these in-line with the client-go pkg versions our current supported Argo CD version uses)

**Comment by the-technat** [comment]
Don't forget to regenerate the docs! Since you changed the schema for `ApplicationSet` you need to make sure the docs are updated. A `make generate` will do that for you.

**Comment by KyriosGN0** [comment]
@the-technat ran make generate, thx!

File: go.mod (hunk 1 above)
Comment (KyriosGN0) [reply go.mod]: reverted back

**Comment by the-technat** [comment]
@KyriosGN0 it looks like the tests fail for 2.9. I'm pretty sure that's because templatePatch support was added in 2.10. Glancing at the code it looks though as if you have added logic to not use this feature if the server is 2.9. Could you reverify this?

In my opinion since it's a feature supported in the last three minor versions we shouldn't have to spend too much time trying to get the tests green on 2.9 which is unsupported anyways. Instead we might want to wait for #472 which will remove 2.9 from the CI anyways. This just as a hint to make it easier for you ;)

**Comment by KyriosGN0** [comment]
@the-technat yeah i tried and it failed, will take a look in a couple of days and will test locally, thx!

**Comment by KyriosGN0** [comment]
@the-technat should work now

**Comment by the-technat** [comment]
I see, thanks @KyriosGN0 for your work!

Will do some manual tests this weekend as you suggested and give it a final review...

**Comment by KyriosGN0** [comment]
hey @the-technat did you have sometime to review my PR ? (did some basic testing myself and it worked)

**Comment by the-technat** [comment]
Sorry for the late reply and thanks for the bump @KyriosGN0. I just reviewed the PR and approved it + rebased the branch. If the tests pass it will be automatically merged soon.

Thanks for your contribution!

**Pull request merged**